diff -rNU3 dist.orig/ld/emulparams/armelf_minix.sh dist/ld/emulparams/armelf_minix.sh
--- dist.orig/ld/emulparams/armelf_minix.sh	2015-12-20 16:45:57.000000000 +0100
+++ dist/ld/emulparams/armelf_minix.sh	2016-01-13 13:21:27.000000000 +0100
@@ -3,13 +3,19 @@
 OUTPUT_FORMAT="elf32-littlearm"
 MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
 COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
-
-DATA_START_SYMBOLS='PROVIDE (__data_start = .);';
+TEXT_START_ADDR=0x00010000
 
 # Dynamic libraries support
-GENERATE_SHLIB_SCRIPT=yes
 TARGET2_TYPE=got-rel
 
-GENERATE_PIE_SCRIPT=yes
-
 unset EMBEDDED
+unset DATA_START_SYMBOLS
+unset STACK_ADDR
+
+# Use the ARM ABI-compliant exception-handling sections.
+OTHER_READONLY_SECTIONS="
+  .ARM.extab ${RELOCATING-0} : { *(.ARM.extab${RELOCATING+* .gnu.linkonce.armextab.*}) }
+  ${RELOCATING+ PROVIDE_HIDDEN (__exidx_start = .); }
+  .ARM.exidx ${RELOCATING-0} : { *(.ARM.exidx${RELOCATING+* .gnu.linkonce.armexidx.*}) }
+  ${RELOCATING+ PROVIDE_HIDDEN (__exidx_end = .); }"
+
